### PR TITLE
purego: fix callback crash for arm64

### PR DESCRIPTION
**Pull request by erdichen**
crosscall2 clobbers FP in the frame record. Use NOFRAME flag for callbackasm1 to disable the assembler generated epilogue that restores the corrupted FP. Save and restore FP manually instead.

### CODE REVIEW

**Comment by hajimehoshi** [comment]
Can we add a test which crashes without the fix?

**Comment by TotallyGamerJet** [comment]
I did run this PR with Oto and Ebitengine and all the tests passed which is good. I do agree that a test that fails without this should be added.

File: callback_test.go
Comment (hajimehoshi) [inline callback_test.go:1]: Isn't it possible to remove the build tag by compiling the C program on the fly (`os/exec`)?

File: callback_test.go
Comment (TotallyGamerJet) [inline callback_test.go:36]: This could be `string(unsafe.Slice(cstr, n))`

File: callback_test.go
Comment (erdichen) [reply callback_test.go]: Done

File: callback_test.go
Comment (erdichen) [reply callback_test.go]: Done. Since `unsafe.String` requires 1.20, I falsely assumed the same for `unsafe.Slice`.

File: callback_test.go
Comment (hajimehoshi) [inline callback_test.go:1]: Add a copy right comment

```
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: 2023 The Ebitengine Authors
```

Comment (hajimehoshi) [inline callback_test.go:1]: Does this test only for Darwin? What about Linux?

File: libcbtest/Makefile
Comment (hajimehoshi) [inline libcbtest/Makefile:2]: We no longer need this file

File: libcbtest/callback.c
Comment (hajimehoshi) [inline libcbtest/callback.c:1]: Add a copy right comment

```
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: 2023 The Ebitengine Authors
```

File: callback_test.go
Comment (hajimehoshi) [inline callback_test.go:25]: `filepath.Join("libctest", "callback.c")`

File: callback_test.go
Comment (hajimehoshi) [inline callback_test.go:61]: running `go env CC` and trimming the result by `strings.TrimSpace` should be fine. We don't need parsing JSON.

File: libcbtest/callback.c
Comment (hajimehoshi) [inline libcbtest/callback.c:36]: Do we need manipulating nodes in order to reproduce the issue? Or would just using a mutex be fine?

**Review by hajimehoshi — COMMENTED** [review]
Thank you!

File: libcbtest/callback.c
Comment (hajimehoshi) [inline libcbtest/callback.c:1]: We need `//go:build ignore` for this file, as this file is unexpectedly included to compile the test.

Comment (hajimehoshi) [reply libcbtest/callback.c:1]: Never mind, the directory was separated.

File: callback_test.go
Comment (erdichen) [reply callback_test.go:4]: Done

Comment (erdichen) [reply callback_test.go:4]: Linux does not support callback. I don't have arm64 Windows.

File: libcbtest/Makefile
Comment (erdichen) [reply libcbtest/Makefile]: Done

File: libcbtest/callback.c
Comment (erdichen) [reply libcbtest/callback.c:4]: Done

File: callback_test.go
Comment (erdichen) [reply callback_test.go]: Done

File: callback_test.go
Comment (erdichen) [reply callback_test.go]: Done

File: libcbtest/callback.c
Comment (erdichen) [reply libcbtest/callback.c:39]: The crash happens after the Go callback returns and the C code access local variables because the frame pointer is incorrect. It was strange when the thread start function is compiled by cgo and is part of the main Go binary, it did not crash. I checked that FP was still overwritten but did not crash. FP was replaced by LR since the crosscall2 frame overlaps by one machine word, so it may just be a coincidence. Access a local variable that points to heap memory is more likely to trigger the crash.

File: callback_test.go
Comment (TotallyGamerJet) [reply callback_test.go:4]: Windows doesn't go through this so just Darwin is fine

File: callback_test.go
Comment (TotallyGamerJet) [inline callback_test.go:57]: Is this supposed to be spawning a goroutine? Since you have WaitGroup

Comment (erdichen) [reply callback_test.go:57]: It is waiting for the async callback from the C pthread.

File: libcbtest/callback.c
Comment (hajimehoshi) [reply libcbtest/callback.c:39]: If we can get the minimized test to reproduce the issue, this would be the best. I'm still not sure the algorithm for the nodes and/or printf outputs. If these are likely needed, please leave a comment to explain why we need these code.

File: callback_test.go
Comment (hajimehoshi) [inline callback_test.go:76]: This ignores stderr. CombinedOutput seems fine

Comment (erdichen) [reply callback_test.go]: Done. Also log the output.

File: libcbtest/callback.c
Comment (erdichen) [reply libcbtest/callback.c]: Turns out just checking a local variable against a known value is good enough. Just need to make sure to disable optimization so the C compiler does not optimize out the local variable.

File: libcbtest/callback.c
Comment (hajimehoshi) [inline libcbtest/callback.c:5]: What about just `callCallback`? This function itself doesn't call printf.

File: libcbtest/callback.c
Comment (hajimehoshi) [inline libcbtest/callback.c:6]: What is this?

File: libcbtest/callback.c
Comment (erdichen) [reply libcbtest/callback.c]: Done

File: libcbtest/callback.c
Comment (erdichen) [reply libcbtest/callback.c:7]: Added comment. If the FP is corrupted, the local variable on the stack will have a different value.

**Review by hajimehoshi — APPROVED** [review]
LGTM, thank you very much!

I would like to wait for @TotallyGamerJet's review.

**Comment by TotallyGamerJet** [comment]
Does callback.c need a license stub? If not, LGTM.

**Pull request merged**
